File: Two Best Non-Overlapping Events.py
# https://leetcode.com/problems/two-best-non-overlapping-events/?envType=daily-question&envId=2024-12-08
# Two Best Non-Overlapping Events

class Solution:
    def maxTwoEvents(self, events: List[List[int]]) -> int:
        
        events.sort(key=lambda x: x[1])
    
        max_value = 0
        max_single_event = []
        end_times = []

        for _, end, value in events:
            max_value = max(max_value, value)
            max_single_event.append(max_value)
            end_times.append(end)

        result = 0
        for start, end, value in events:
            idx = bisect_right(end_times, start - 1) - 1
            if idx >= 0:
                result = max(result, value + max_single_event[idx])
            else:
                result = max(result, value)
        
        return result

File: Two Best Non-Overlapping Events.js
// https://leetcode.com/problems/two-best-non-overlapping-events/?envType=daily-question&envId=2025-12-23
// Two Best Non-Overlapping Events

/**
 * @param {number[][]} events
 * @return {number}
 */
var maxTwoEvents = function(events) {
    events.sort((a, b) => a[1] - b[1]);

    let maxValue = 0;
    const maxSingleEvent = [];
    const endTimes = [];

    for (const [, end, value] of events) {
        maxValue = Math.max(maxValue, value);
        maxSingleEvent.push(maxValue);
        endTimes.push(end);
    }

    let result = 0;

    function bisectRight(arr, target) {
        let left = 0, right = arr.length;
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (arr[mid] <= target) left = mid + 1;
            else right = mid;
        }
        return left;
    }

    for (const [start, , value] of events) {
        const idx = bisectRight(endTimes, start - 1) - 1;

        if (idx >= 0) {
            result = Math.max(result, value + maxSingleEvent[idx]);
        } else {
            result = Math.max(result, value);
        }
    }

    return result;
};